moveit_core/online_signal_smoothing/src/butterworth_filter.cpp

```diff
@@ -0,0 +1,106 @@
+#include <moveit/online_signal_smoothing/butterworth_filter.h>
+
+namespace online_signal_smoothing
+{
+namespace
+{
+constexpr double EPSILON = 1e-9;
+}
+
+ButterworthFilter::ButterworthFilter(double low_pass_filter_coeff)
+  : previous_measurements_{ 0., 0. }
+  , previous_filtered_measurement_(0.)
+  , scale_term_(1. / (1. + low_pass_filter_coeff))
+  , feedback_term_(1. - low_pass_filter_coeff)
+{
+  // guarantee this doesn't change because the logic below depends on this length implicity
+  static_assert(ButterworthFilter::FILTER_LENGTH == 2,
+                "online_signal_smoothing::ButterworthFilter::FILTER_LENGTH should be 2");
+
+  if (std::isinf(feedback_term_))
+    throw std::length_error("online_signal_smoothing::ButterworthFilter: infinite feedback_term_");
+
+  if (std::isinf(scale_term_))
+    throw std::length_error("online_signal_smoothing::ButterworthFilter: infinite scale_term_");
+
+  if (low_pass_filter_coeff < 1)
+    throw std::length_error(
+        "online_signal_smoothing::ButterworthFilter: Filter coefficient < 1. makes the lowpass filter unstable");
+
+  if (std::abs(feedback_term_) < EPSILON)
+    throw std::length_error(
+        "online_signal_smoothing::ButterworthFilter: Filter coefficient value resulted in feedback term of 0");
+}
+
+double ButterworthFilter::filter(double new_measurement)
+{
+  // Push in the new measurement
+  previous_measurements_[1] = previous_measurements_[0];
+  previous_measurements_[0] = new_measurement;
+
+  double new_filtered_measurement = scale_term_ * (previous_measurements_[1] + previous_measurements_[0] -
```

Comment:
Directly store in `previous_filtered_measurement_` and return that one?

Reply:
ah, that's a nice little optimization. And here I thought this code was quite streamlined already.